scripts/compose-user-provided-service.sh

```diff
@@ -0,0 +1,171 @@
+SCRIPT=$0
+COMPOSE_API_URL='https://api.compose.io/2016-07'
+
+# shellcheck disable=SC1090
+source "${SCRIPT_DIR}/common.sh"
+
+usage() {
+  cat <<EOF
+  Usage:
+    $SCRIPT [-n <name>] -o <orgname> -s <space[,space,...]> -t <type>
+
+$SCRIPT will manually create an instance
+of defined type with the use of compose API and assign it to the specific
+organisation as a CloudFoundry User Provided Service.
+
+Requirements:
+
+ * You must login with the cf client with an "admin" user.
+ * You must have the compose_api_token environment variable set.
+
+Where:
+  -n <name>    Name of the service that should be created. If flag is not set
+               a default value will be generated.
+
+  -o <orgname> Organisation the service should be assigned to upon creation.
+
+  -s <space>   Comma separated space names the service should be assigned to
+               upon creation.
+
+  -t <type>    Service type to be created. Supported:
+               mongodb, redis, postgresql, elastic_search, rethink, rabbitmq,
+               etcd, scylla, mysql, disque
+EOF
+exit 1
+}
+
+get_cluster_data(){
+  compose_cluster_data=$(curl -s -X GET "${COMPOSE_API_URL}/clusters" \
+    -H "Authorization: Bearer ${COMPOSE_API_TOKEN}" \
+    -H "Content-Type: application/json" \
+    2>/dev/null | jq -r ._embedded.clusters[0])
+
+  compose_cluster_id=$(echo "${compose_cluster_data}" | jq -r .id)
+  compose_account_id=$(echo "${compose_cluster_data}" | jq -r .account_id)
+}
+
+prepare_name() {
+  id=$(date +%s | md5)
+  deployment_name=$(echo "alpha-${TYPE}-${ORG}-${id}" | cut -c1-50)
+}
+
+provision_instance() {
+  local body
+  body=$(printf '{
+    "deployment": {
+      "name": "%s",
+      "account_id": "%s",
+      "cluster_id": "%s",
+      "type": "%s"
+    }
+  }' "${SERVICE_NAME:-$deployment_name}" "${compose_account_id}" "${compose_cluster_id}" "${TYPE}")
+
+  uri=$(curl -s -X POST "${COMPOSE_API_URL}/deployments" \
+      -H "Authorization: Bearer ${COMPOSE_API_TOKEN}" \
+      -H "Content-Type: application/json; charset=utf-8" \
+      -d "${body}" | jq -r .connection_strings.direct[0])
+
+  pattern='^(([a-z]{3,5})://)?((([^:\/]+)(:([^@\/]*))?@)?([^:\/?]+)(:([0-9]+))?)(\/[^?]*)?(\?[^#]*)?(#.*)?$'
+  [[ "$uri" =~ $pattern ]] || return 1;
+
+  uri=${BASH_REMATCH[0]}
+  uri_user=${BASH_REMATCH[5]}
+  uri_password=${BASH_REMATCH[7]}
+  uri_host=${BASH_REMATCH[8]}
+  uri_port=${BASH_REMATCH[10]}
+}
+
+create_user_provided_service() {
+  local body
+  body=$(printf '{
+    "username": "%s",
+    "password": "%s",
+    "host": "%s",
+    "port": "%s",
+    "uri": "%s"
+  }' "${uri_user}" "${uri_password}" "${uri_host}" "${uri_port}" "${uri}")
+
+  IFS=',';
+  for i in $(printf "%s" "${SPACE}")
+  do
+    info "Targeting organisation '${ORG}' and space '${i}'."
+
+    cf target -o "${ORG}" -s "${i}"
+
+    warning "Please bare in mind, if prior to this message you'll encounter an error, you will need to manually deprovision the instance: '${deployment_name}'"
```

Comment:
I'm picking on tiny things, but the latter was a little confusing:
s/bare/bear/
s/prior/post/
After that, time to merge :)

Reply:
English language hard. Thanks! :D

Reply:
:) I've been told english is a nightmare!